File: apps/backend/ModuAgent/orchestration/communication/message_bus.py
from __future__ import annotations

import asyncio
import logging
from collections import defaultdict
from contextlib import contextmanager
from typing import Any, Callable, Coroutine, Dict, Iterator, List, Optional

from .protocol import AgentEvent, EventPriority

logger = logging.getLogger(__name__)

EventHandler = Callable[[AgentEvent], Coroutine[Any, Any, None]]


class Subscription:
    def __init__(
        self,
        handler: EventHandler,
        domain: Optional[str] = None,
        action: Optional[str] = None,
        priority_filter: Optional[EventPriority] = None,
    ):
        self.handler = handler
        self.domain = domain
        self.action = action
        self.priority_filter = priority_filter

    def matches(self, event: AgentEvent) -> bool:
        if self.domain and self.domain != event.domain:
            return False
        if self.action and self.action != event.action:
            return False
        if self.priority_filter and self.priority_filter != event.priority:
            return False
        return True


class EventBus:
    def __init__(self):
        self._subscriptions: List[Subscription] = []
        self._domain_index: Dict[str, List[Subscription]] = defaultdict(list)
        # P2-2: 移除内存事件日志 _event_log/_max_log_size 及保护它的 _lock，
        # 事件持久化统一由 PersistentEventLog 订阅处理（避免双份内存开销与职责重叠）。
        # 在单线程 asyncio 事件循环中，subscribe/publish 的同步段不会被 await 之间打断，
        # 因此无需额外锁保护订阅集合。

    def subscribe(
        self,
        handler: EventHandler,
        domain: Optional[str] = None,
        action: Optional[str] = None,
        priority_filter: Optional[EventPriority] = None,
    ) -> Callable[[], None]:
        sub = Subscription(
            handler=handler,
            domain=domain,
            action=action,
            priority_filter=priority_filter,
        )
        self._subscriptions.append(sub)
        if domain:
            self._domain_index[domain].append(sub)

        def unsubscribe():
            if sub in self._subscriptions:
                self._subscriptions.remove(sub)
            if domain and sub in self._domain_index.get(domain, []):
                self._domain_index[domain].remove(sub)

        return unsubscribe

    async def publish(self, event: AgentEvent) -> None:
        # P2-2: 不再追加到内存 _event_log，事件持久化由 PersistentEventLog 订阅处理。
        matched = self._domain_index.get(event.domain, self._subscriptions)
        if not matched:
            matched = self._subscriptions

        tasks = []
        for sub in matched:
            if sub.matches(event):
                tasks.append(self._safe_invoke(sub.handler, event))

        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)

    async def _safe_invoke(self, handler: EventHandler, event: AgentEvent) -> None:
        try:
            await handler(event)
        except Exception as e:
            logger.error(
                "Event handler error: event_id=%s domain=%s action=%s error=%s",
                event.event_id,
                event.domain,
                event.action,
                str(e),
            )

    async def request(
        self,
        event: AgentEvent,
        timeout_ms: int = 5000,
    ) -> Optional[AgentEvent]:
        response_future: asyncio.Future[AgentEvent] = asyncio.get_event_loop().create_future()
        request_id = event.event_id

        async def response_handler(resp_event: AgentEvent):
            if resp_event.metadata.get("request_id") == request_id and not response_future.done():
                response_future.set_result(resp_event)

        unsub = self.subscribe(
            response_handler,
            domain=event.domain,
            action=f"{event.action}_response",
        )

        await self.publish(event)

        try:
            return await asyncio.wait_for(
                response_future,
                timeout=timeout_ms / 1000.0,
            )
        except asyncio.TimeoutError:
            logger.warning("Request timeout: event_id=%s domain=%s", request_id, event.domain)
            return None
        finally:
            unsub()


class PersistentEventLog:
    """事件日志持久化处理器（P1：事件追踪信息完整化）。

    订阅 EventBus 的所有事件，将事件以 JSONL 格式写入文件。
    支持按 domain 过滤、文件轮转。

    设计原则：
    - 文件写入异步，不阻塞主流程
    - 文件不可用时降级为内存日志
    - 支持最大文件大小限制和轮转
    """

    def __init__(
        self,
        log_file_path: str,
        max_file_size_mb: float = 10.0,
        domains: Optional[List[str]] = None,
    ) -> None:
        """初始化持久化事件日志。

        Args:
            log_file_path: 日志文件路径（JSONL 格式）
            max_file_size_mb: 单文件最大大小（MB），超过后轮转
            domains: 仅持久化指定 domain 的事件（None=全部）
        """
        self._log_file_path = log_file_path
        self._max_file_size = int(max_file_size_mb * 1024 * 1024)
        self._domains = set(domains) if domains else None
        self._enabled = False
        self._write_queue: asyncio.Queue = asyncio.Queue()
        self._writer_task: Optional[asyncio.Task] = None

    async def start(self, event_bus: EventBus) -> None:
        """启动持久化日志：订阅事件并启动写入任务。"""
        import os
        log_dir = os.path.dirname(self._log_file_path)
        if log_dir and not os.path.exists(log_dir):
            try:
                os.makedirs(log_dir, exist_ok=True)
            except OSError as e:
                logger.warning("Cannot create log directory %s: %s", log_dir, str(e))
                return

        self._enabled = True
        event_bus.subscribe(self._on_event)
        self._writer_task = asyncio.create_task(self._writer_loop())
        logger.info("PersistentEventLog started: %s", self._log_file_path)

    async def stop(self) -> None:
        """停止持久化日志。"""
        self._enabled = False
        if self._writer_task:
            self._writer_task.cancel()
            try:
                await self._writer_task
            except asyncio.CancelledError:
                pass
            self._writer_task = None

    async def _on_event(self, event: AgentEvent) -> None:
        """事件回调：将事件放入写入队列。"""
        if not self._enabled:
            return
        if self._domains and event.domain not in self._domains:
            return
        await self._write_queue.put(event)

    async def _writer_loop(self) -> None:
        """后台写入循环：从队列取事件写入文件。"""
        import json
        import os

        while self._enabled:
            try:
                event = await asyncio.wait_for(self._write_queue.get(), timeout=1.0)
            except asyncio.TimeoutError:
                continue
            except asyncio.CancelledError:
                break

            try:
                # 文件轮转检查
                if os.path.exists(self._log_file_path):
                    file_size = os.path.getsize(self._log_file_path)
                    if file_size > self._max_file_size:
                        self._rotate_log()

                # 序列化事件
                event_dict = {
                    "event_id": event.event_id,
                    "timestamp": event.timestamp,
                    "trace_id": event.trace_id,
                    "session_id": event.session_id,
                    "user_id": event.user_id,
                    "domain": event.domain,
                    "action": event.action,
                    "priority": event.priority,
                    "metadata": event.metadata,
                }
                line = json.dumps(event_dict, ensure_ascii=False) + "\n"

                with open(self._log_file_path, "a", encoding="utf-8") as f:
                    f.write(line)
            except Exception as e:
                logger.warning("Failed to write event log: %s", str(e))

    def _rotate_log(self) -> None:
        """日志文件轮转：重命名为 .1 后缀。"""
        import os
        rotated_path = self._log_file_path + ".1"
        try:
            if os.path.exists(rotated_path):
                os.unlink(rotated_path)
            os.rename(self._log_file_path, rotated_path)
            logger.info("Event log rotated: %s → %s", self._log_file_path, rotated_path)
        except OSError as e:
            logger.warning("Log rotation failed: %s", str(e))


_event_bus: Optional[EventBus] = None


def get_event_bus(override: Optional[EventBus] = None) -> EventBus:
    """获取全局 EventBus 单例。

    P2-1: 新增 `override` 参数用于测试隔离。
    生产代码不应使用此参数；测试在 teardown 中应调用 `reset_event_bus()` 清理。

    Args:
        override: 测试时注入的实例。若提供，将替换全局单例并返回。

    Returns:
        全局 EventBus 实例
    """
    global _event_bus
    if override is not None:
        _event_bus = override
    if _event_bus is None:
        _event_bus = EventBus()
    return _event_bus


def reset_event_bus() -> None:
    """重置全局 event_bus 单例（测试清理用）。"""
    global _event_bus
    _event_bus = None


@contextmanager
def override_event_bus(event_bus: EventBus) -> Iterator[EventBus]:
    """P2-1: 测试用上下文管理器——临时替换全局 event_bus 单例，退出时自动恢复。"""
    global _event_bus
    old = _event_bus
    _event_bus = event_bus
    try:
        yield _event_bus
    finally:
        _event_bus = old

File: packages/modu-agent/src/orchestration/communication/message-bus.ts
// 对应 Python: orchestration/communication/message_bus.py
// EventBus + 全局单例 get_event_bus + PersistentEventLog
import fs from 'fs'
import { mkdir, appendFile, stat, rename, unlink } from 'fs/promises'
import path from 'path'

import type { AgentEvent, EventPriority } from './protocol.js'

const logger = {
  info: (msg: string, ...args: any[]) => console.info(`[event-bus] ${msg}`, ...args),
  warning: (msg: string, ...args: any[]) => console.warn(`[event-bus] ${msg}`, ...args),
  error: (msg: string, ...args: any[]) => console.error(`[event-bus] ${msg}`, ...args),
  debug: (msg: string, ...args: any[]) => console.debug(`[event-bus] ${msg}`, ...args),
}

export type EventHandler = (event: AgentEvent) => Promise<void> | void

// ============================================================
// Subscription
// ============================================================

export interface SubscriptionOptions {
  handler: EventHandler
  domain?: string | null
  action?: string | null
  priority_filter?: EventPriority | null
}

export class Subscription {
  handler: EventHandler
  domain: string | null
  action: string | null
  priority_filter: EventPriority | null

  constructor(opts: SubscriptionOptions) {
    this.handler = opts.handler
    this.domain = opts.domain ?? null
    this.action = opts.action ?? null
    this.priority_filter = opts.priority_filter ?? null
  }

  matches(event: AgentEvent): boolean {
    if (this.domain && this.domain !== event.domain) {
      return false
    }
    if (this.action && this.action !== event.action) {
      return false
    }
    if (this.priority_filter && this.priority_filter !== event.priority) {
      return false
    }
    return true
  }
}

// ============================================================
// EventBus
// ============================================================

export class EventBus {
  private _subscriptions: Subscription[] = []
  private _domainIndex: Map<string, Subscription[]> = new Map()

  subscribe(
    handler: EventHandler,
    domain?: string | null,
    action?: string | null,
    priority_filter?: EventPriority | null,
  ): () => void {
    const sub = new Subscription({ handler, domain, action, priority_filter })
    this._subscriptions.push(sub)
    if (domain) {
      let list = this._domainIndex.get(domain)
      if (!list) {
        list = []
        this._domainIndex.set(domain, list)
      }
      list.push(sub)
    }

    return () => {
      const idx = this._subscriptions.indexOf(sub)
      if (idx >= 0) {
        this._subscriptions.splice(idx, 1)
      }
      if (domain) {
        const list = this._domainIndex.get(domain)
        if (list) {
          const i2 = list.indexOf(sub)
          if (i2 >= 0) {
            list.splice(i2, 1)
          }
        }
      }
    }
  }

  async publish(event: AgentEvent): Promise<void> {
    let matched = this._domainIndex.get(event.domain)
    if (!matched || matched.length === 0) {
      matched = this._subscriptions
    }

    const tasks: Promise<void>[] = []
    for (const sub of matched) {
      if (sub.matches(event)) {
        tasks.push(this._safeInvoke(sub.handler, event))
      }
    }

    if (tasks.length > 0) {
      await Promise.allSettled(tasks)
    }
  }

  private async _safeInvoke(handler: EventHandler, event: AgentEvent): Promise<void> {
    try {
      await handler(event)
    } catch (e) {
      logger.error(
        'Event handler error: event_id=%s domain=%s action=%s error=%s',
        event.event_id, event.domain, event.action, String(e),
      )
    }
  }

  async request(event: AgentEvent, timeoutMs = 5000): Promise<AgentEvent | null> {
    let resolveFn!: (val: AgentEvent | null) => void
    const responsePromise = new Promise<AgentEvent | null>((resolve) => {
      resolveFn = resolve
    })
    let done = false
    const request_id = event.event_id

    const responseHandler = (respEvent: AgentEvent): void => {
      if (respEvent.metadata.request_id === request_id && !done) {
        done = true
        resolveFn(respEvent)
      }
    }

    const unsub = this.subscribe(
      responseHandler,
      event.domain,
      `${event.action}_response`,
    )

    await this.publish(event)

    const timeoutPromise = new Promise<AgentEvent | null>((resolve) => {
      setTimeout(() => {
        if (!done) {
          done = true
          resolve(null)
        }
      }, timeoutMs)
    })

    try {
      return await Promise.race([responsePromise, timeoutPromise])
    } finally {
      unsub()
    }
  }
}

// ============================================================
// PersistentEventLog（对应 Python PersistentEventLog）
// ============================================================

export class PersistentEventLog {
  private _log_file_path: string
  private _max_file_size: number
  private _domains: Set<string> | null
  private _enabled = false
  private _write_queue: AgentEvent[] = []
  private _writer_running = false

  constructor(
    log_file_path: string,
    max_file_size_mb = 10.0,
    domains?: string[] | null,
  ) {
    this._log_file_path = log_file_path
    this._max_file_size = Math.floor(max_file_size_mb * 1024 * 1024)
    this._domains = domains ? new Set(domains) : null
  }

  async start(event_bus: EventBus): Promise<void> {
    const log_dir = path.dirname(this._log_file_path)
    if (log_dir && !fs.existsSync(log_dir)) {
      try {
        await mkdir(log_dir, { recursive: true })
      } catch (e) {
        logger.warning('Cannot create log directory %s: %s', log_dir, String(e))
        return
      }
    }

    this._enabled = true
    event_bus.subscribe(this._on_event.bind(this))
    this._writer_running = true
    this._writerLoop()
    logger.info('PersistentEventLog started: %s', this._log_file_path)
  }

  async stop(): Promise<void> {
    this._enabled = false
    this._writer_running = false
    // 等待队列排空
    while (this._write_queue.length > 0) {
      await new Promise((r) => setTimeout(r, 50))
    }
  }

  private async _on_event(event: AgentEvent): Promise<void> {
    if (!this._enabled) {
      return
    }
    if (this._domains && !this._domains.has(event.domain)) {
      return
    }
    this._write_queue.push(event)
  }

  private async _writerLoop(): Promise<void> {
    while (this._writer_running) {
      if (this._write_queue.length === 0) {
        await new Promise((r) => setTimeout(r, 1000))
        continue
      }
      const event = this._write_queue.shift()!
      try {
        if (fs.existsSync(this._log_file_path)) {
          const stats = await stat(this._log_file_path)
          if (stats.size > this._max_file_size) {
            await this._rotateLog()
          }
        }
        const event_dict = {
          event_id: event.event_id,
          timestamp: event.timestamp.toISOString(),
          trace_id: event.trace_id,
          session_id: event.session_id,
          user_id: event.user_id,
          domain: event.domain,
          action: event.action,
          priority: event.priority,
          metadata: event.metadata,
        }
        const line = JSON.stringify(event_dict) + '\n'
        await appendFile(this._log_file_path, line, 'utf-8')
      } catch (e) {
        logger.warning('Failed to write event log: %s', String(e))
      }
    }
  }

  private async _rotateLog(): Promise<void> {
    const rotated_path = this._log_file_path + '.1'
    try {
      if (fs.existsSync(rotated_path)) {
        await unlink(rotated_path)
      }
      await rename(this._log_file_path, rotated_path)
      logger.info('Event log rotated: %s → %s', this._log_file_path, rotated_path)
    } catch (e) {
      logger.warning('Log rotation failed: %s', String(e))
    }
  }
}

// ============================================================
// 全局单例（对应 Python get_event_bus / reset_event_bus）
// ============================================================

let _event_bus: EventBus | null = null

export function get_event_bus(override?: EventBus | null): EventBus {
  if (override !== undefined && override !== null) {
    _event_bus = override
  }
  if (_event_bus === null) {
    _event_bus = new EventBus()
  }
  return _event_bus
}

export function reset_event_bus(): void {
  _event_bus = null
}

export function override_event_bus(event_bus: EventBus): { restore: () => void } {
  const old = _event_bus
  _event_bus = event_bus
  return {
    restore: () => {
      _event_bus = old
    },
  }
}
